jax/_src/sourcemap.py

```diff
@@ -0,0 +1,215 @@
+
+# Each segment is made up of 1, 4 or 5 variable-length fields
+# (see https://tc39.es/source-map/#mappings-structure).
+Segment = Union[tuple[int], tuple[int, int, int, int], tuple[int, int, int, int, int]]
+
+@dataclass(frozen=True)
+class SourceMap:
+    version: int
+    # file: str
+    # source_root: str
+    sources: list[str]
+    sources_content: list[str]
+    names: list[str]
+    mappings: list[list[Segment]]
+
+
+VLQ_SIGN_MASK = 0x01
+VLQ_MORE_MASK = 0x20
+VLQ_VALUE_MASK = 0x1f
+VLQ_VALUE_BITWIDTH = 5
+VLQ_ALPHABET = (
+    list(range(ord("A"), ord("Z")+1)) +
+    list(range(ord("a"), ord("z")+1)) +
+    list(range(ord("0"), ord("9")+1)) +
+    [ord("+"), ord("/")]
+)
+
+
+def make_vlq_decode_table():
+    lookup = {c: d for d, c in enumerate(VLQ_ALPHABET)}
+    return [lookup.get(i, None) for i in range(256)]
+
+VLQ_DECODE_TABLE = make_vlq_decode_table()
+
+
+def decode_vlq(enc: Iterable[int]) -> int:
+    """Decode a Base-64-VLQ into an integer."""
+    enc_iter = iter(enc)
+    d = VLQ_DECODE_TABLE[next(enc_iter)]
+    sign = bool(d & VLQ_SIGN_MASK)
+    value = (d & VLQ_VALUE_MASK) >> 1
+    # Compensate for first quantum containing sign as LSB:
+    shift = -1
+
+    while d & VLQ_MORE_MASK:
+        shift += VLQ_VALUE_BITWIDTH
+        d = VLQ_DECODE_TABLE[next(enc_iter)]
+        value |= (d & VLQ_VALUE_MASK) << shift
+
+    return -value if sign else value
+
+
+def encode_vlq(value: int) -> bytes:
+    """Encode an integer into a Base-64-VLQ."""
+    # Move sign to LSB
+    value = ((-value) << 1 | 1) if value < 0 else value << 1
+    buf = []
+
+    while True:
+        d = value & VLQ_VALUE_MASK
+        value >>= VLQ_VALUE_BITWIDTH
+        more = value > 0
+        if more:
+            d |= VLQ_MORE_MASK
+        buf.append(VLQ_ALPHABET[d])
+        if not more:
+            break
+    return bytes(buf)
+
+
+def decode_segment(enc: Iterable[int]) -> Segment:
+    """Decode a sequence of VLQs into a segment."""
+    enc_iter = iter(enc)
+    col = decode_vlq(enc_iter)
+    try:
+        source = decode_vlq(enc_iter)
+    except StopIteration:
+        # Stopping here is fine (1-segment).
+        return (col,)
+    source_line = decode_vlq(enc_iter)
+    source_col = decode_vlq(enc_iter)
+    try:
+        name = decode_vlq(enc_iter)
+    except StopIteration:
+        # Stopping here is fine too (4-segment).
+        return col, source, source_line, source_col
+    # (5-segment)
+    return col, source, source_line, source_col, name
+
+
+def encode_segment(seg: Segment) -> bytes:
+    """Encode a segment into a sequence of VLQs."""
+    return b"".join(encode_vlq(value) for value in seg)
+
+
+def deserialize_mappings(mappings_str: str) -> list[list[Segment]]:
+    """Decode a string of TC39 mapping data into a list of segments per line."""
+    mappings_bytes = bytes(mappings_str, encoding="ascii")
+    return [
+        list(map(decode_segment, mapping.split(b","))) if mapping else []
```

Comment:
Nit: wouldn't `list(map(decode_segment, mapping.split(b",")))` work just fine for an empty `mapping` without the if-else?

Reply:
Unfortunately not, since `b"".split(b",") == [b""]`.

For example given
```
mappings_bytes = b"A;"
```
I want to produce
```
[[decode_segment(b"A")], []]
```
rather than
```
[[decode_segment(b"A")], [decode_segment(b"")]]
```

Reply:
Right! I missed that `b"".split(b",")` is not `[]` but `[b""]`.